<!DOCTYPE html>
<html>
<!--custom header -->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width initial-scale=1">

  <title> BADays (Bioinformatics Awareness Days)</title>
  <meta name="description" content="Bioinformatics Awareness Days (BAD days)">


  <link rel="shortcut icon" href="http://localhost:4000/BAD_days/images/dna.png" >

  <!-- Style sheets -->
  <!--Material bootstrap -->
  <link href="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.3.0/css/mdb.min.css" rel="stylesheet" medial ="all">

  <!-- Font awesome -->
  <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css" rel="stylesheet" media="all">

  <!-- Boostrap-->
  <link rel="stylesheet" href="/BAD_days/css/main.css">


  <!-- URL -->
  <link rel="canonical" href="http://localhost:4000/BAD_days/Multiple Hyp.html">


  <!-- The various JS scripts used in the website -->
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
  <script src="https://use.fontawesome.com/ff95ae1185.js"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"
  integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa"
  crossorigin="anonymous"></script>

<!-- add after bootstrap.min.js -->
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.3.0/js/mdb.min.js"></script>
<link src="/BAD_days/js/custom-scripts.js">


<!-- Loading mathjax macro -->
<!-- Load mathjax -->
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <!-- MathJax configuration -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ['$','$'], ["\\(","\\)"] ],
            displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
            processEscapes: true,
            processEnvironments: true
        },
        // Center justify equations in code and markdown cells. Elsewhere
        // we use CSS to left justify single line equations in code cells.
        displayAlign: 'center',
        "HTML-CSS": {
            styles: {'.MathJax_Display': {"margin": 0}},
            linebreaks: { automatic: true }
        }
    });
    </script>

</head>


<!-- scroller for TOC-->
<body data-spy="scroll" data-target="#toc">

<!-- sidebar-->
  <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox">

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <p>Bioinformatics Awareness Days (BAD days)</p>

    <p> Developed by Luisa Cutillo and the <a href='http://rse.shef.ac.uk'> RSE team
    </a> at the University of Sheffield</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/BAD_days/">Home</a>

    

    
    
      
        
        <!-- Note you need to prepend the site.baseurl always-->
          <a class="sidebar-nav-item"
          href="/BAD_days/1-outline/">Day 1</a>
        
      
    
      
        
        <!-- Note you need to prepend the site.baseurl always-->
          <a class="sidebar-nav-item"
          href="/BAD_days/2-outline/">Day 2</a>
        
      
    
      
        
        <!-- Note you need to prepend the site.baseurl always-->
          <a class="sidebar-nav-item"
          href="/BAD_days/3-outline/">Day 3</a>
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
    
      
        
      
    
      
    


    <a class="sidebar-nav-item" href="https://github.com/trallard/BAD_days">GitHub repository</a>
  </nav>

  <div class="sidebar-item">
    <p>
      The University of Sheffield &copy; 2017. CC-BY
    </p>
  </div>
</div>


<!--main content needs to be in a wrap for
when the sidebar is displayed-->
  <div class='wrap'>

<!-- navbar -->
  <!--Top navigation bar-->

    <nav class="navbar navbar-inverse">
      <div class="container-fluid">

        <a class="navbar-brand" href="/BAD_days/index.html"> </a>

        </div>
      </div>
    </nav>

</div>
<div class='wrap'>
  <div class="container">
    <div class="row">

<!--- Main content--->
      <div class="col-sm-8">
        <main>
          <h1 id="bad-day-1-multiple-hypothesis-testing-on-golub">BAD Day 1: Multiple hypothesis testing on Golub</h1>

<p>In this tutorial we will do multiple hypothesis testing on the previsouly
studied <code class="highlighter-rouge">Golub</code> data set.</p>

<ul>
  <li><strong>Null Hypothesis</strong> $H_0$ states the assumption to be tested. We begin by
assuming that the null Hypothesis is True</li>
  <li><strong>Alternative Hypothesis</strong> $H_1$ is the opposite of the null hypothesis</li>
</ul>

<h2 id="loading-the-dataset">1. Loading the dataset</h2>

<p><br /></p>
<font color="#00bcd4"> In [1]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">require</span><span class="p">(</span><span class="n">multtest</span><span class="p">)</span><span class="w">
</span><span class="n">data</span><span class="p">(</span><span class="n">golub</span><span class="p">)</span></code></pre></figure>

<p>Gene expression data (3051 genes and 38 tumor mRNA samples) from the leukemia
microarray study of Golub et al. (1999). Pre-processing was done as described in
Dudoit et al. (2002). The R code for pre-processing is available in the file
../doc/golub.R.</p>

<p><br /></p>
<font color="#00bcd4"> In [2]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">golub.expr</span><span class="o">&lt;-</span><span class="n">golub</span><span class="w">
</span><span class="n">row.names</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">)</span><span class="o">=</span><span class="n">golub.gnames</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span></code></pre></figure>

<p>Numeric vector indicating the tumor class, 27 acute lymphoblastic leukemia (ALL)
cases (code 0) and 11 acute myeloid leukemia (AML) cases (code 1).</p>

<p><br /></p>
<font color="#00bcd4"> In [3]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">colnames</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">)</span><span class="o">=</span><span class="w"> </span><span class="n">golub.cl</span></code></pre></figure>

<p>Setting the sample sizes</p>

<p><br /></p>
<font color="#00bcd4"> In [4]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">n.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">27</span><span class="w">
</span><span class="n">n.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">11</span><span class="w">
</span><span class="n">cancer.type</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s2">"ALL"</span><span class="p">,</span><span class="w"> </span><span class="n">n.ALL</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"AML"</span><span class="p">,</span><span class="w"> </span><span class="n">n.AML</span><span class="p">))</span></code></pre></figure>

<p>Adding the cancer type to the column name, for the display</p>

<p><br /></p>
<font color="#00bcd4"> In [5]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">colnames</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">cancer.type</span></code></pre></figure>

<h2 id="ttest-with-a-single-gene">2. t.test with a single gene</h2>

<p><br /></p>
<font color="#00bcd4"> In [6]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">g</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">347</span></code></pre></figure>

<p>Alternatively, you can select a gene randomly</p>

<p><br /></p>
<font color="#00bcd4"> In [7]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1">## g &lt;- sample(1:nrow(golub.expr),1)
</span><span class="n">g.profile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.vector</span><span class="p">(</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">[</span><span class="n">g</span><span class="p">,]))</span></code></pre></figure>

<p>Draw a barplot wiht color-coded cancer type</p>

<p><br /></p>
<font color="#00bcd4"> In [9]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">plot.col</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'ALL'</span><span class="o">=</span><span class="s1">'mediumpurple'</span><span class="p">,</span><span class="w"> </span><span class="s1">'AML'</span><span class="o">=</span><span class="s1">'midnightblue'</span><span class="p">)</span><span class="w">

</span><span class="n">barplot</span><span class="p">(</span><span class="n">g.profile</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="s2">"Golub (1999), gene"</span><span class="p">,</span><span class="w"> </span><span class="n">g</span><span class="p">),</span><span class="w">
        </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plot.col</span><span class="p">[</span><span class="n">cancer.type</span><span class="p">],</span><span class="w"> </span><span class="n">border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'white'</span><span class="p">)</span><span class="w">

</span><span class="n">legend</span><span class="p">(</span><span class="s1">'topright'</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"ALL"</span><span class="p">,</span><span class="s2">"AML"</span><span class="p">),</span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plot.col</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ALL"</span><span class="p">,</span><span class="s2">"AML"</span><span class="p">)],</span><span class="w">
       </span><span class="n">pch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="n">bty</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"o"</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>

<p><img src="http://localhost:4000/BAD_days/notebooks/Multiple%20Hyp_files/Multiple%20Hyp_16_0.png" alt="png" /></p>

<p>Separate data in two vectors</p>

<p><br /></p>
<font color="#00bcd4"> In [10]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sample.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">g.profile</span><span class="p">[</span><span class="n">cancer.type</span><span class="o">==</span><span class="s2">"ALL"</span><span class="p">]</span><span class="w">
</span><span class="n">sample.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">g.profile</span><span class="p">[</span><span class="n">cancer.type</span><span class="o">==</span><span class="s2">"AML"</span><span class="p">]</span></code></pre></figure>

<p>Compute <strong>manually</strong> the t test paramenters (not necessary, just to practice!)</p>

<p>Estimate the population means</p>

<p><br /></p>
<font color="#00bcd4"> In [11]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">mean.est.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">)</span><span class="w">
</span><span class="n">mean.est.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean</span><span class="p">(</span><span class="n">sample.AML</span><span class="p">)</span></code></pre></figure>

<p>Compute the sample sd</p>

<p>Remember the <code class="highlighter-rouge">sd( )</code> function automatically computes the estimate corrected with
sqrt(n-1)</p>

<p><br /></p>
<font color="#00bcd4"> In [12]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sample.sd.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">((</span><span class="n">n.ALL</span><span class="m">-1</span><span class="p">)</span><span class="o">/</span><span class="n">n.ALL</span><span class="p">)</span><span class="w">
</span><span class="n">sample.sd.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.AML</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">((</span><span class="n">n.AML</span><span class="m">-1</span><span class="p">)</span><span class="o">/</span><span class="n">n.AML</span><span class="p">)</span></code></pre></figure>

<p>Estimate the population standard deviation</p>

<p><br /></p>
<font color="#00bcd4"> In [13]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sd.est.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">)</span><span class="w">
</span><span class="n">sd.est.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.AML</span><span class="p">)</span></code></pre></figure>

<p>Estimate the standard errors on the mean</p>

<p><br /></p>
<font color="#00bcd4"> In [14]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sd.err.est.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="n">n.ALL</span><span class="p">)</span><span class="w">
</span><span class="n">sd.err.est.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.AML</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="n">n.AML</span><span class="p">)</span></code></pre></figure>

<p>Estimate the standard deviation of the difference between the two means
according to Student’s formula</p>

<p><br /></p>
<font color="#00bcd4"> In [15]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">diff.sd.est</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">((</span><span class="n">n.ALL</span><span class="o">*</span><span class="n">sample.sd.ALL</span><span class="o">^</span><span class="m">2</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">n.AML</span><span class="o">*</span><span class="n">sample.sd.AML</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">/</span><span class="n">n.ALL</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="o">/</span><span class="n">n.AML</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="p">(</span><span class="n">n.ALL</span><span class="o">+</span><span class="n">n.AML</span><span class="m">-2</span><span class="p">))</span></code></pre></figure>

<p>Compute <code class="highlighter-rouge">t.obs</code></p>

<p><br /></p>
<font color="#00bcd4"> In [16]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">mean.est.ALL</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">mean.est.AML</span><span class="p">)</span><span class="w">
</span><span class="n">t.obs.Student</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">d</span><span class="o">/</span><span class="n">diff.sd.est</span></code></pre></figure>

<p>Compute the P- vale.
Since we are performing the two-tail test, the single-tail probability has to be
multiplied by 3 in order to obtai the alpha risk</p>

<p><br /></p>
<font color="#00bcd4"> In [17]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">P.val.Student</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">pt</span><span class="p">(</span><span class="n">q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">t.obs.Student</span><span class="p">,</span><span class="w"> </span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n.ALL</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">n.AML</span><span class="m">-2</span><span class="p">,</span><span class="w"> </span><span class="n">lower.tail</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span></code></pre></figure>

<h3 id="this-is-what-you-should-be-doing-fast">This is what you should be doing… FAST!</h3>

<h2 id="apply-the-student-fischer-t-test-this-assumes-that-the-two-populations">3. Apply the Student-Fischer t-test (this assumes that the two populations</h2>
<p>have equal variance).</p>

<p><br /></p>
<font color="#00bcd4"> In [18]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">t.student</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">t.test</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">,</span><span class="n">sample.AML</span><span class="p">,</span><span class="w"> </span><span class="n">var.equal</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">print</span><span class="p">(</span><span class="n">t.student</span><span class="p">)</span></code></pre></figure>

<h2 id="apply-the-welch-t-test-this-does-not-assume-that-the-two-populations-have">4. Apply the Welch t-test (this does not assume that the two populations have</h2>
<p>equal variance)</p>

<p><br /></p>
<font color="#00bcd4"> In [19]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">t.welch</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">t.test</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">,</span><span class="n">sample.AML</span><span class="p">,</span><span class="w"> </span><span class="n">var.equal</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">print</span><span class="p">(</span><span class="n">t.welch</span><span class="p">)</span></code></pre></figure>

<p><strong>We want to test all the genes, so we could just loop over what was written
before … (note it can take some time)</strong></p>

<p><br /></p>
<font color="#00bcd4"> In [20]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">t.statistics</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vector</span><span class="p">()</span><span class="w">
</span><span class="n">P.values</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vector</span><span class="p">()</span><span class="w">
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">g</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">))</span><span class="w"> </span><span class="p">{</span><span class="w">
  </span><span class="c1">#  print(paste("Testing gene", g))
</span><span class="w">  </span><span class="n">g.profile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.vector</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">[</span><span class="n">g</span><span class="p">,])</span><span class="w">
  </span><span class="n">sample.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">g.profile</span><span class="p">[</span><span class="n">cancer.type</span><span class="o">==</span><span class="s2">"ALL"</span><span class="p">]</span><span class="w">
  </span><span class="n">sample.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">g.profile</span><span class="p">[</span><span class="n">cancer.type</span><span class="o">==</span><span class="s2">"AML"</span><span class="p">]</span><span class="w">
  </span><span class="n">t</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">t.test</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">,</span><span class="n">sample.AML</span><span class="p">)</span><span class="w">
  </span><span class="n">t.statistics</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">append</span><span class="p">(</span><span class="n">t.statistics</span><span class="p">,</span><span class="w"> </span><span class="n">t</span><span class="o">$</span><span class="n">statistic</span><span class="p">)</span><span class="w">
  </span><span class="n">P.values</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">append</span><span class="p">(</span><span class="n">P.values</span><span class="p">,</span><span class="w"> </span><span class="n">t</span><span class="o">$</span><span class="n">p.value</span><span class="p">)</span><span class="w">
</span><span class="p">}</span></code></pre></figure>

<p><br /></p>
<font color="#00bcd4"> In [21]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="w"> </span><span class="n">print</span><span class="p">(</span><span class="n">P.values</span><span class="p">)</span></code></pre></figure>

<h4 id="for-a-more-efficient-way-you-can-use-apply">For a more efficient way, you can use apply</h4>

<p><br /></p>
<font color="#00bcd4"> In [22]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">Data</span><span class="o">=</span><span class="n">cbind</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">,</span><span class="n">P.values</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">Data</span><span class="p">)[</span><span class="m">39</span><span class="p">]</span><span class="o">=</span><span class="s1">'Raw.p'</span></code></pre></figure>

<p>Order data by pvalue</p>

<p><br /></p>
<font color="#00bcd4"> In [23]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">Data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Data</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">Data</span><span class="p">[,</span><span class="s1">'Raw.p'</span><span class="p">]),]</span></code></pre></figure>

<p>Perform p-value adjustments and add to the data frame</p>

<p><br /></p>
<font color="#00bcd4"> In [24]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">Datadf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">Data</span><span class="p">)</span></code></pre></figure>

<p><br /></p>
<font color="#00bcd4"> In [25]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">Datadf</span><span class="o">$</span><span class="n">Bonferroni</span><span class="w"> </span><span class="o">=</span><span class="w">
  </span><span class="n">p.adjust</span><span class="p">(</span><span class="n">Datadf</span><span class="o">$</span><span class="n">Raw.p</span><span class="p">,</span><span class="w">
           </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bonferroni"</span><span class="p">)</span><span class="w">

</span><span class="n">Datadf</span><span class="o">$</span><span class="n">BH</span><span class="w"> </span><span class="o">=</span><span class="w">
  </span><span class="n">p.adjust</span><span class="p">(</span><span class="n">Datadf</span><span class="o">$</span><span class="n">Raw.p</span><span class="p">,</span><span class="w">
           </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">)</span><span class="w">

</span><span class="n">Datadf</span><span class="o">$</span><span class="n">Holm</span><span class="w"> </span><span class="o">=</span><span class="w">
  </span><span class="n">p.adjust</span><span class="p">(</span><span class="n">Datadf</span><span class="o">$</span><span class="w"> </span><span class="n">Raw.p</span><span class="p">,</span><span class="w">
           </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"holm"</span><span class="p">)</span><span class="w">

</span><span class="n">Datadf</span><span class="o">$</span><span class="n">Hochberg</span><span class="w"> </span><span class="o">=</span><span class="w">
  </span><span class="n">p.adjust</span><span class="p">(</span><span class="n">Datadf</span><span class="o">$</span><span class="w"> </span><span class="n">Raw.p</span><span class="p">,</span><span class="w">
           </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"hochberg"</span><span class="p">)</span><span class="w">

</span><span class="n">Datadf</span><span class="o">$</span><span class="n">Hommel</span><span class="w"> </span><span class="o">=</span><span class="w">
  </span><span class="n">p.adjust</span><span class="p">(</span><span class="n">Datadf</span><span class="o">$</span><span class="w"> </span><span class="n">Raw.p</span><span class="p">,</span><span class="w">
           </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"hommel"</span><span class="p">)</span><span class="w">

</span><span class="n">Datadf</span><span class="o">$</span><span class="n">BY</span><span class="w"> </span><span class="o">=</span><span class="w">
  </span><span class="n">p.adjust</span><span class="p">(</span><span class="n">Datadf</span><span class="o">$</span><span class="w"> </span><span class="n">Raw.p</span><span class="p">,</span><span class="w">
           </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BY"</span><span class="p">)</span></code></pre></figure>

<h2 id="plotting-the-obtained-data">5. Plotting the obtained data</h2>

<p><br /></p>
<font color="#00bcd4"> In [26]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">X</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Datadf</span><span class="o">$</span><span class="n">Raw.p</span><span class="w">
</span><span class="n">Y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="n">Datadf</span><span class="o">$</span><span class="n">Bonferroni</span><span class="p">,</span><span class="w">
          </span><span class="n">Datadf</span><span class="o">$</span><span class="n">BH</span><span class="p">,</span><span class="w">
          </span><span class="n">Datadf</span><span class="o">$</span><span class="n">Holm</span><span class="p">,</span><span class="w">
          </span><span class="n">Datadf</span><span class="o">$</span><span class="n">Hochberg</span><span class="p">,</span><span class="w">
          </span><span class="n">Datadf</span><span class="o">$</span><span class="n">Hommel</span><span class="p">,</span><span class="w">
          </span><span class="n">Datadf</span><span class="o">$</span><span class="n">BY</span><span class="p">)</span></code></pre></figure>

<p><br /></p>
<font color="#00bcd4"> In [27]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">matplot</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="w"> </span><span class="n">Y</span><span class="p">,</span><span class="w">
        </span><span class="n">xlab</span><span class="o">=</span><span class="s2">"Raw p-value"</span><span class="p">,</span><span class="w">
        </span><span class="n">ylab</span><span class="o">=</span><span class="s2">"Adjusted p-value"</span><span class="p">,</span><span class="w">
        </span><span class="n">type</span><span class="o">=</span><span class="s2">"l"</span><span class="p">,</span><span class="w">
        </span><span class="n">asp</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w">
        </span><span class="n">col</span><span class="o">=</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span><span class="w">
        </span><span class="n">lty</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w">
        </span><span class="n">lwd</span><span class="o">=</span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">legend</span><span class="p">(</span><span class="s1">'bottomright'</span><span class="p">,</span><span class="w">
       </span><span class="n">legend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Bonferroni"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Holm"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Hochberg"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Hommel"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BY"</span><span class="p">),</span><span class="w">
       </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span><span class="w">
       </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w">    
       </span><span class="n">pch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">16</span><span class="p">)</span><span class="w">

</span><span class="n">abline</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w">
       </span><span class="n">col</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w">
       </span><span class="n">lty</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w">
       </span><span class="n">lwd</span><span class="o">=</span><span class="m">1</span><span class="p">)</span></code></pre></figure>

<p><img src="http://localhost:4000/BAD_days/notebooks/Multiple%20Hyp_files/Multiple%20Hyp_50_0.png" alt="png" /></p>

        </main>
      </div>

      <div class="col-sm-1">
      </div>
      <!-- TOC sidebar, which will move to the top on a small screen -->
      <div class="col-sm-3 container-toc">
        <nav id="toc" data-spy="affix" data-toggle="toc"></nav>
      </div>
    </div>
  </div>
</div>

<label for="sidebar-checkbox" class="sidebar-toggle"></label>

</body>


</html>
