<!DOCTYPE html>
<html>
<!--custom header -->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width initial-scale=1">

  <title> BADays (Bioinformatics Awareness Days)</title>
  <meta name="description" content="Bioinformatics Awareness Days (BAD days)">


  <link rel="shortcut icon" href="http://localhost:4000/BAD_days/images/dna.png" >

  <!-- Style sheets -->
  <!--Material bootstrap -->
  <link href="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.3.0/css/mdb.min.css" rel="stylesheet" medial ="all">

  <!-- Font awesome -->
  <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css" rel="stylesheet" media="all">

  <!-- Boostrap-->
  <link rel="stylesheet" href="/BAD_days/css/main.css">


  <!-- URL -->
  <link rel="canonical" href="http://localhost:4000/BAD_days/Ttest.html">


  <!-- The various JS scripts used in the website -->
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
  <script src="https://use.fontawesome.com/ff95ae1185.js"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"
  integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa"
  crossorigin="anonymous"></script>

<!-- add after bootstrap.min.js -->
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.3.0/js/mdb.min.js"></script>
<link src="/BAD_days/js/custom-scripts.js">


<!-- Loading mathjax macro -->
<!-- Load mathjax -->
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <!-- MathJax configuration -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ['$','$'], ["\\(","\\)"] ],
            displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
            processEscapes: true,
            processEnvironments: true
        },
        // Center justify equations in code and markdown cells. Elsewhere
        // we use CSS to left justify single line equations in code cells.
        displayAlign: 'center',
        "HTML-CSS": {
            styles: {'.MathJax_Display': {"margin": 0}},
            linebreaks: { automatic: true }
        }
    });
    </script>

</head>


<!-- scroller for TOC-->
<body data-spy="scroll" data-target="#toc">

<!-- sidebar-->
  <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox">

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <p>Bioinformatics Awareness Days (BAD days)</p>

    <p> Developed by Luisa Cutillo and the <a href='http://rse.shef.ac.uk'> RSE team
    </a> at the University of Sheffield</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/BAD_days/">Home</a>

    

    
    
      
        
        <!-- Note you need to prepend the site.baseurl always-->
          <a class="sidebar-nav-item"
          href="/BAD_days/1-outline/">Day 1</a>
        
      
    
      
        
        <!-- Note you need to prepend the site.baseurl always-->
          <a class="sidebar-nav-item"
          href="/BAD_days/2-outline/">Day 2</a>
        
      
    
      
        
        <!-- Note you need to prepend the site.baseurl always-->
          <a class="sidebar-nav-item"
          href="/BAD_days/3-outline/">Day 3</a>
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
        
      
    
      
    
      
        
      
    
      
    


    <a class="sidebar-nav-item" href="https://github.com/trallard/BAD_days">GitHub repository</a>
  </nav>

  <div class="sidebar-item">
    <p>
      The University of Sheffield &copy; 2017. CC-BY
    </p>
  </div>
</div>


<!--main content needs to be in a wrap for
when the sidebar is displayed-->
  <div class='wrap'>

<!-- navbar -->
  <!--Top navigation bar-->

    <nav class="navbar navbar-inverse">
      <div class="container-fluid">

        <a class="navbar-brand" href="/BAD_days/index.html"> </a>

        </div>
      </div>
    </nav>

</div>
<div class='wrap'>
  <div class="container">
    <div class="row">

<!--- Main content--->
      <div class="col-sm-8">
        <main>
          <h1 id="bad-day-1--t-test">BAD Day 1 : T-test</h1>

<h2 id="loading-the-data-set">1. Loading the data set</h2>

<p><br /></p>
<font color="#00bcd4"> In [1]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">require</span><span class="p">(</span><span class="n">multtest</span><span class="p">)</span><span class="w">
</span><span class="n">data</span><span class="p">(</span><span class="n">golub</span><span class="p">)</span></code></pre></figure>

<p>We will be usig the Gene Expression dataset from Golub et al (1999).</p>

<p>Gene expression data (3051 genes and 38 tumor mRNA samples) from the leukemia
microarray study of Golub et al. (1999). Pre-processing was done as described in
Dudoit et al. (2002). The R code for pre-processing is available in the file
../doc/golub.R.
Source: Golub et al. (1999). Molecular classification of cancer: class discovery
and class prediction by gene expression monitoring, Science, Vol. 286:531-537.
http://www-genome.wi.mit.edu/MPR/ .</p>

<p><br /></p>
<font color="#00bcd4"> In [2]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">golub.expr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">golub</span><span class="w">

</span><span class="c1"># preliminar view of the data
</span><span class="n">head</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">)</span></code></pre></figure>

<table class="table-responsive table-striped">
<tbody>
	<tr><td>-1.45769</td><td>-1.39420</td><td>-1.42779</td><td>-1.40715</td><td>-1.42668</td><td>-1.21719</td><td>-1.37386</td><td>-1.36832</td><td>-1.47649</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td>-1.26183</td><td>-1.44434</td><td> 1.10147</td><td>-1.34158</td><td>-1.22961</td><td>-0.75919</td><td>0.84905 </td><td>-0.66465</td></tr>
	<tr><td>-0.75161</td><td>-1.26278</td><td>-0.09052</td><td>-0.99596</td><td>-1.24245</td><td>-0.69242</td><td>-1.37386</td><td>-0.50803</td><td>-1.04533</td><td>-0.81257</td><td>⋯       </td><td>-1.08902</td><td>-1.05094</td><td>-1.26183</td><td>-1.25918</td><td> 0.97813</td><td>-0.79357</td><td>-1.22961</td><td>-0.71792</td><td>0.45127 </td><td>-0.45804</td></tr>
	<tr><td> 0.45695</td><td>-0.09654</td><td> 0.90325</td><td>-0.07194</td><td> 0.03232</td><td> 0.09713</td><td>-0.11978</td><td> 0.23381</td><td> 0.23987</td><td> 0.44201</td><td>⋯       </td><td>-0.43377</td><td>-0.10823</td><td>-0.29385</td><td> 0.05067</td><td> 1.69430</td><td>-0.12472</td><td> 0.04609</td><td> 0.24347</td><td>0.90774 </td><td> 0.46509</td></tr>
	<tr><td> 3.13533</td><td> 0.21415</td><td> 2.08754</td><td> 2.23467</td><td> 0.93811</td><td> 2.24089</td><td> 3.36576</td><td> 1.97859</td><td> 2.66468</td><td>-1.21583</td><td>⋯       </td><td> 0.29598</td><td>-1.29865</td><td> 2.76869</td><td> 2.08960</td><td> 0.70003</td><td> 0.13854</td><td> 1.75908</td><td> 0.06151</td><td>1.30297 </td><td> 0.58186</td></tr>
	<tr><td> 2.76569</td><td>-1.27045</td><td> 1.60433</td><td> 1.53182</td><td> 1.63728</td><td> 1.85697</td><td> 3.01847</td><td> 1.12853</td><td> 2.17016</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td> 2.00518</td><td> 1.17454</td><td>-1.47218</td><td>-1.34158</td><td> 1.55086</td><td>-1.18107</td><td>1.01596 </td><td> 0.15788</td></tr>
	<tr><td> 2.64342</td><td> 1.01416</td><td> 1.70477</td><td> 1.63845</td><td>-0.36075</td><td> 1.73451</td><td> 3.36576</td><td> 0.96870</td><td> 2.72368</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td> 1.73780</td><td> 0.89347</td><td>-0.52883</td><td>-1.22168</td><td> 0.90832</td><td>-1.39906</td><td>0.51266 </td><td> 1.36249</td></tr>
</tbody>
</table>

<p><code class="highlighter-rouge">golub.names</code> is a matrix containing the names of the 3051
 genes contained in <code class="highlighter-rouge">golub</code>. The three columns correspond to:
<code class="highlighter-rouge">index, ID and name</code></p>

<p><br /></p>
<font color="#00bcd4"> In [3]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">row.names</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">golub.gnames</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span></code></pre></figure>

<p><code class="highlighter-rouge">golub.cl</code> is a numeric vector indicating the tumor class, 27 acute
lymphoblastic leukemia (ALL) cases (code 0) and 11 acute myeloid leukemia (AML)
cases (code 1).</p>

<p><br /></p>
<font color="#00bcd4"> In [4]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">colnames</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">golub.cl</span></code></pre></figure>

<p>Now we need to set the sample sizes</p>

<p><br /></p>
<font color="#00bcd4"> In [5]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">n.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">27</span><span class="w">
</span><span class="n">n.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">11</span><span class="w">

</span><span class="n">cancer.type</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s1">'ALL'</span><span class="p">,</span><span class="w"> </span><span class="n">n.ALL</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s1">'AML'</span><span class="p">,</span><span class="w"> </span><span class="n">n.AML</span><span class="p">))</span></code></pre></figure>

<p>Adding the cancer type to the column name, for the display</p>

<p><br /></p>
<font color="#00bcd4"> In [6]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">colnames</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cancer.type</span></code></pre></figure>

<p><br /></p>
<font color="#00bcd4"> In [7]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">head</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">)</span></code></pre></figure>

<table class="table-responsive table-striped">
<thead><tr><th></th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">ALL</th><th scope="col">⋯</th><th scope="col">AML</th><th scope="col">AML</th><th scope="col">AML</th><th scope="col">AML</th><th scope="col">AML</th><th scope="col">AML</th><th scope="col">AML</th><th scope="col">AML</th><th scope="col">AML</th><th scope="col">AML</th></tr></thead>
<tbody>
	<tr><th scope="row">AFFX-HUMISGF3A/M97935_MA_at</th><td>-1.45769</td><td>-1.39420</td><td>-1.42779</td><td>-1.40715</td><td>-1.42668</td><td>-1.21719</td><td>-1.37386</td><td>-1.36832</td><td>-1.47649</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td>-1.26183</td><td>-1.44434</td><td> 1.10147</td><td>-1.34158</td><td>-1.22961</td><td>-0.75919</td><td>0.84905 </td><td>-0.66465</td></tr>
	<tr><th scope="row">AFFX-HUMISGF3A/M97935_MB_at</th><td>-0.75161</td><td>-1.26278</td><td>-0.09052</td><td>-0.99596</td><td>-1.24245</td><td>-0.69242</td><td>-1.37386</td><td>-0.50803</td><td>-1.04533</td><td>-0.81257</td><td>⋯       </td><td>-1.08902</td><td>-1.05094</td><td>-1.26183</td><td>-1.25918</td><td> 0.97813</td><td>-0.79357</td><td>-1.22961</td><td>-0.71792</td><td>0.45127 </td><td>-0.45804</td></tr>
	<tr><th scope="row">AFFX-HUMISGF3A/M97935_3_at</th><td> 0.45695</td><td>-0.09654</td><td> 0.90325</td><td>-0.07194</td><td> 0.03232</td><td> 0.09713</td><td>-0.11978</td><td> 0.23381</td><td> 0.23987</td><td> 0.44201</td><td>⋯       </td><td>-0.43377</td><td>-0.10823</td><td>-0.29385</td><td> 0.05067</td><td> 1.69430</td><td>-0.12472</td><td> 0.04609</td><td> 0.24347</td><td>0.90774 </td><td> 0.46509</td></tr>
	<tr><th scope="row">AFFX-HUMRGE/M10098_5_at</th><td> 3.13533</td><td> 0.21415</td><td> 2.08754</td><td> 2.23467</td><td> 0.93811</td><td> 2.24089</td><td> 3.36576</td><td> 1.97859</td><td> 2.66468</td><td>-1.21583</td><td>⋯       </td><td> 0.29598</td><td>-1.29865</td><td> 2.76869</td><td> 2.08960</td><td> 0.70003</td><td> 0.13854</td><td> 1.75908</td><td> 0.06151</td><td>1.30297 </td><td> 0.58186</td></tr>
	<tr><th scope="row">AFFX-HUMRGE/M10098_M_at</th><td> 2.76569</td><td>-1.27045</td><td> 1.60433</td><td> 1.53182</td><td> 1.63728</td><td> 1.85697</td><td> 3.01847</td><td> 1.12853</td><td> 2.17016</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td> 2.00518</td><td> 1.17454</td><td>-1.47218</td><td>-1.34158</td><td> 1.55086</td><td>-1.18107</td><td>1.01596 </td><td> 0.15788</td></tr>
	<tr><th scope="row">AFFX-HUMRGE/M10098_3_at</th><td> 2.64342</td><td> 1.01416</td><td> 1.70477</td><td> 1.63845</td><td>-0.36075</td><td> 1.73451</td><td> 3.36576</td><td> 0.96870</td><td> 2.72368</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td> 1.73780</td><td> 0.89347</td><td>-0.52883</td><td>-1.22168</td><td> 0.90832</td><td>-1.39906</td><td>0.51266 </td><td> 1.36249</td></tr>
</tbody>
</table>

<h2 id="ttest-with-a-single-gene-step-by-step">2. t.test with a single gene (step-by-step)</h2>

<p>We will select the gene 347 for this example</p>

<p><br /></p>
<font color="#00bcd4"> In [8]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">g</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">347</span></code></pre></figure>

<p>Alternatively, you can select a gene randomly</p>

<p><br /></p>
<font color="#00bcd4"> In [9]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># g &lt;- sample(1:nrow(golub.expr),1)
</span><span class="n">g.profile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.vector</span><span class="p">(</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">[</span><span class="n">g</span><span class="p">,]))</span></code></pre></figure>

<p>Draw a barplot with color-coded cancer type</p>

<p><br /></p>
<font color="#00bcd4"> In [10]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">plot.col</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'ALL'</span><span class="o">=</span><span class="s1">'midnightblue'</span><span class="p">,</span><span class="w"> </span><span class="s1">'AML'</span><span class="o">=</span><span class="s1">'mediumpurple'</span><span class="p">)</span><span class="w">

</span><span class="n">barplot</span><span class="p">(</span><span class="n">g.profile</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="s2">"Golub (1999), gene"</span><span class="p">,</span><span class="w"> </span><span class="n">g</span><span class="p">),</span><span class="w">
        </span><span class="n">col</span><span class="o">=</span><span class="n">plot.col</span><span class="p">[</span><span class="n">cancer.type</span><span class="p">])</span><span class="w">
</span><span class="n">legend</span><span class="p">(</span><span class="s1">'topright'</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"ALL"</span><span class="p">,</span><span class="s2">"AML"</span><span class="p">),</span><span class="n">col</span><span class="o">=</span><span class="n">plot.col</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ALL"</span><span class="p">,</span><span class="s2">"AML"</span><span class="p">)],</span><span class="w">
       </span><span class="n">pch</span><span class="o">=</span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="n">bty</span><span class="o">=</span><span class="s2">"o"</span><span class="p">,</span><span class="w"> </span><span class="n">bg</span><span class="o">=</span><span class="s1">'white'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>

<p><img src="http://localhost:4000/BAD_days/notebooks/Ttest_files/Ttest_20_0.png" alt="png" /></p>

<p>Separating the data in two vectors</p>

<p><br /></p>
<font color="#00bcd4"> In [11]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sample.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">g.profile</span><span class="p">[</span><span class="n">cancer.type</span><span class="o">==</span><span class="s2">"ALL"</span><span class="p">]</span><span class="w">
</span><span class="n">sample.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">g.profile</span><span class="p">[</span><span class="n">cancer.type</span><span class="o">==</span><span class="s2">"AML"</span><span class="p">]</span></code></pre></figure>

<p>**Compute manually ** the t test parameters (not necessary, just to practice!)</p>

<p>Estimate the population means</p>

<p><br /></p>
<font color="#00bcd4"> In [12]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">mean.est.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">)</span><span class="w">
</span><span class="n">mean.est.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean</span><span class="p">(</span><span class="n">sample.AML</span><span class="p">)</span></code></pre></figure>

<p>Compute the sample standard deviations
The sd() function automatically computes the estimate corrected with sqrt(n-1)</p>

<p><br /></p>
<font color="#00bcd4"> In [13]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sample.sd.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">((</span><span class="n">n.ALL</span><span class="m">-1</span><span class="p">)</span><span class="o">/</span><span class="n">n.ALL</span><span class="p">)</span><span class="w">
</span><span class="n">sample.sd.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.AML</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">((</span><span class="n">n.AML</span><span class="m">-1</span><span class="p">)</span><span class="o">/</span><span class="n">n.AML</span><span class="p">)</span></code></pre></figure>

<p>Estimate the population standard deviation</p>

<p><br /></p>
<font color="#00bcd4"> In [14]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sd.est.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">)</span><span class="w">
</span><span class="n">sd.est.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.AML</span><span class="p">)</span></code></pre></figure>

<p>Estimate the standard deviations on the means</p>

<p><br /></p>
<font color="#00bcd4"> In [15]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">sd.err.est.ALL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.ALL</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="n">n.ALL</span><span class="p">)</span><span class="w">
</span><span class="n">sd.err.est.AML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">sample.AML</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="n">n.AML</span><span class="p">)</span></code></pre></figure>

<p>Estimate the standard deviation of the difference between two means, according
to Student’s formula</p>

<p><br /></p>
<font color="#00bcd4"> In [16]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">diff.sd.est</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">((</span><span class="n">n.ALL</span><span class="o">*</span><span class="n">sample.sd.ALL</span><span class="o">^</span><span class="m">2</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">n.AML</span><span class="o">*</span><span class="n">sample.sd.AML</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">/</span><span class="n">n.ALL</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="o">/</span><span class="n">n.AML</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="p">(</span><span class="n">n.ALL</span><span class="o">+</span><span class="n">n.AML</span><span class="m">-2</span><span class="p">))</span></code></pre></figure>

<p>Compute t.obs</p>

<p><br /></p>
<font color="#00bcd4"> In [17]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">mean.est.ALL</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">mean.est.AML</span><span class="p">)</span><span class="w">
</span><span class="n">t.obs.Student</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">d</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">diff.sd.est</span></code></pre></figure>

<p>Compute the P- vale.
Since we are performing the two-tail test, the single-tail probability has to be
multiplied by 3 in order to obtai the alpha risk</p>

<p><br /></p>
<font color="#00bcd4"> In [18]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">P.val.Student</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">pt</span><span class="p">(</span><span class="n">q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">t.obs.Student</span><span class="p">,</span><span class="w"> </span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n.ALL</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">n.AML</span><span class="m">-2</span><span class="p">,</span><span class="w"> </span><span class="n">lower.tail</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span></code></pre></figure>

<h2 id="t-test-the-fast-way">3. T-test the fast way</h2>

<p>**This is what you should be doing… **</p>

<p>Suppose that the gene expression data from two groups of patients
(experimental) are availaeble and that the that the hypothesis is about the
difference
between the population means $\mu_1$ and $\mu_2$.</p>

<p>Thus $H_0: \mu_1 = \mu_2$ is to be tested against  $H_0: \mu_1 \neq \mu_2$. If
the gene expression data are $\left( x_1…x_n \right)$ and
$\left(y_1…y_n\right)$ for the first and second group, respectively.</p>

<p>With the mean and variances of the first and the second groups being $\bar{x}$,
$\bar{y}$ , $s_1^2$,$s_2^2$.</p>

<p>Then the t-statistic can be formulated as:</p>

<script type="math/tex; mode=display">t =
\frac{(\bar{x}-\bar{y})-(\mu_1-\mu_2)}{\sqrt{\frac{s_1^2}{n}+\frac{s_2^2}{m}}}</script>

<p><strong>Note</strong>: the t-value is large if the difference between $\bar{x}$ and $\bar{y}$
is large and the standard deviations $s_1$ and $s_2$ are small. This corresponds
to the <strong>Welch two-sample t-test</strong>.</p>

<h3 id="two-samples-t-test-with-unequal-variances-welch-t-test">3.2 Two-samples t-test with unequal variances (Welch t-test)</h3>

<p>Golub et al. (1999) argue that gene CCND3 Cyclin D3 plays
an important role with respect to discriminating ALL from AML patients. We
previously generated the boxplots for this gene, which suggests that the  <code class="highlighter-rouge">ALL</code>
population means differs from that of the <code class="highlighter-rouge">AML</code>. The null hypothesis of equal
means can be tested by using the appropriate factor and specification <code class="highlighter-rouge">var.equal
= FALSE</code>.</p>

<p><br /></p>
<font color="#00bcd4"> In [22]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1">#t.test(sample.ALL,sample.AML, var.equal=FALSE)
</span><span class="w">
</span><span class="c1"># creating the gol.fac
</span><span class="n">gol.fac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">  </span><span class="n">factor</span><span class="p">(</span><span class="n">golub.cl</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="o">:</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"AML"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ALL"</span><span class="p">))</span><span class="w">
</span><span class="n">t.test</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">gol.fac</span><span class="p">,</span><span class="w"> </span><span class="n">var.equal</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span></code></pre></figure>

<div class="highlighter-rouge"><pre class="highlight"><code>	Welch Two Sample t-test

data:  golub[1042, ] by gol.fac
t = 6.3186, df = 16.118, p-value = 9.871e-06
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 0.8363826 1.6802008
sample estimates:
mean in group AML mean in group ALL
        1.8938826         0.6355909
</code></pre>
</div>

<p>The <em>t</em> value is significantly large, meaning that the two means differ largely
with respect to the corresponding (standard error). As the <em>p</em> value is very
small,the conclusion is to reject the
null-hypothesis of equal means. The data provide strong evidence that the
population means do differ.</p>

<h3 id="two-samples-student-fischer-t-test-samples-with-equal-variances">3.1 Two-samples Student-Fischer t-test (samples with equal variances).</h3>

<p>Suppose the same setup as in the above example. In this case, however, the
variances $\sigma_1^2$ and $\sigma_2^2$ for the two groups are known to be
equal. To test $H_0: \mu_1 = \mu_2$ against $H_1: \mu_1 \neq \mu_2$ there is a
t-test which is based on the pooled sample variance $s_p^2$. The latter is
defined by the weighted sum of the sample variances $s_1^2$ and $s_2^2$ e.g.</p>

<script type="math/tex; mode=display">s_p^2 = \frac{(n-1)s_1^2+(m-1)s_2^2}{n+m-2}</script>

<p>Then the t-value will be</p>

<script type="math/tex; mode=display">t = \frac{\bar{x}-\bar{y}-(\mu_1-\mu_2)}{s_p \sqrt{\frac{1}{n}+\frac{1}{m}}}</script>

<p>The null hypothesis for gene CCND3 Cyclin D3 that the mean of ALL differs from
that of AML patients can be tested by the two-sample t-test using:</p>

<p><br /></p>
<font color="#00bcd4"> In [23]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1">#t.test(sample.ALL,sample.AML, var.equal=TRUE)
</span><span class="w">

</span><span class="n">t.test</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">gol.fac</span><span class="p">,</span><span class="w"> </span><span class="n">var.equal</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span></code></pre></figure>

<div class="highlighter-rouge"><pre class="highlight"><code>	Two Sample t-test

data:  golub[1042, ] by gol.fac
t = 6.7983, df = 36, p-value = 6.046e-08
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 0.8829143 1.6336690
sample estimates:
mean in group AML mean in group ALL
        1.8938826         0.6355909
</code></pre>
</div>

<p>From the p-value $6.046 \times 10^{-8}$, the conclusion is to reject the null
hypothesis
of equal population means. Note that the p-value is slightly smaller than
that of the previous test.</p>

<p><br /></p>
<font color="#00bcd4"> In [None]: </font>


        </main>
      </div>

      <div class="col-sm-1">
      </div>
      <!-- TOC sidebar, which will move to the top on a small screen -->
      <div class="col-sm-3 container-toc">
        <nav id="toc" data-spy="affix" data-toggle="toc"></nav>
      </div>
    </div>
  </div>
</div>

<label for="sidebar-checkbox" class="sidebar-toggle"></label>

</body>


</html>
