<!DOCTYPE html>
<html>
<!--custom header -->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width initial-scale=1">

  <title> BADays (Bioinformatics Awareness Days)</title>
  <meta name="description" content="Bioinformatics Awareness Days (BAD days)">

  <!-- Style sheets -->
  <link href="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.3.0/css/mdb.min.css" rel="stylesheet" medial ="all">
  <link href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css" rel="stylesheet" media="all">

  <link rel="stylesheet" href="/BAD_days/css/main.css">


 <!-- URL -->
  <link rel="canonical" href="http://localhost:4000/BAD_days/Tutorial.html">


  <!-- The various JS scripts used in the website -->
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
  <script src="https://use.fontawesome.com/ff95ae1185.js"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"
  integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa"
  crossorigin="anonymous"></script>

<!-- add after bootstrap.min.js -->
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.3.0/js/mdb.min.js"></script>
<link src="/BAD_days/js/custom-scripts.js">


<!-- Loading mathjax macro -->
<!-- Load mathjax -->
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <!-- MathJax configuration -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ['$','$'], ["\\(","\\)"] ],
            displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
            processEscapes: true,
            processEnvironments: true
        },
        // Center justify equations in code and markdown cells. Elsewhere
        // we use CSS to left justify single line equations in code cells.
        displayAlign: 'center',
        "HTML-CSS": {
            styles: {'.MathJax_Display': {"margin": 0}},
            linebreaks: { automatic: true }
        }
    });
    </script>

</head>


<!-- scroller for TOC-->
<body data-spy="scroll" data-target="#toc">

<!-- sidebar-->
  <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox">

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <p>Bioinformatics Awareness Days (BAD days)</p>

    <p> Developed by Luisa Cutillo and the <a href='http://rse.shef.ac.uk'> RSE team
    </a> at the University of Sheffield</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/BAD_days/">Home</a>

    

    
    
      
        
        <!-- Note you need to prepend the site.baseurl always-->
          <a class="sidebar-nav-item"
          href="/BAD_days/1-outline/">Day 1 outline</a>
        
      
    
      
    
      
        
      
    
      
        
        <!-- Note you need to prepend the site.baseurl always-->
          <a class="sidebar-nav-item"
          href="/BAD_days/about/">Getting started</a>
        
      
    
      
    
      
    


    <a class="sidebar-nav-item" href="https://github.com/trallard/BAD_days">GitHub repository</a>
  </nav>

  <div class="sidebar-item">
    <p>
      The University of Sheffield &copy; 2017. CC-BY
    </p>
  </div>
</div>


<!--main content needs to be in a wrap for
when the sidebar is displayed-->
  <div class='wrap'>

<!-- navbar -->
  <!--Top navigation bar-->

    <nav class="navbar navbar-inverse">
      <div class="container-fluid">

        <a class="navbar-brand" href="/BAD_days/index.html"> </a>

        </div>
      </div>
    </nav>

</div>
<div class='wrap'>
  <div class="container">
    <div class="row">

<!--- Main content--->
      <div class="col-sm-8">
        <main>
          <h1 id="bad-day-1-tutorial">BAD Day 1: Tutorial</h1>

<h1 id="sourceinstall-the-needed-packages">0. Source/install the needed packages</h1>

<p><br /></p>
<font color="#00bcd4"> In [1]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># In case you need to install the packages
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"xlsx"</span><span class="p">)</span><span class="w">
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"gdata"</span><span class="p">)</span><span class="w">
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"ape"</span><span class="p">)</span></code></pre></figure>

<div class="highlighter-rouge"><pre class="highlight"><code>The downloaded binary packages are in
	/var/folders/1q/xdx6qpy905dbx01t7cfv36280000gn/T//RtmpcogYw9/downloaded_packages

The downloaded binary packages are in
	/var/folders/1q/xdx6qpy905dbx01t7cfv36280000gn/T//RtmpcogYw9/downloaded_packages

The downloaded binary packages are in
	/var/folders/1q/xdx6qpy905dbx01t7cfv36280000gn/T//RtmpcogYw9/downloaded_packages
</code></pre>
</div>

<p><br /></p>
<font color="#00bcd4"> In [2]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">source</span><span class="p">(</span><span class="s2">"http://bioconductor.org/biocLite.R"</span><span class="p">);</span><span class="w">
</span><span class="n">biocLite</span><span class="p">(</span><span class="s2">"multtest"</span><span class="p">);</span></code></pre></figure>

<div class="highlighter-rouge"><pre class="highlight"><code>Bioconductor version 3.4 (BiocInstaller 1.24.0), ?biocLite for help
A new version of Bioconductor is available after installing the most recent
  version of R; see http://bioconductor.org/install
BioC_mirror: https://bioconductor.org
Using Bioconductor 3.4 (BiocInstaller 1.24.0), R 3.3.2 (2016-10-31).
Installing package(s) ‘multtest’



The downloaded binary packages are in
	/var/folders/1q/xdx6qpy905dbx01t7cfv36280000gn/T//RtmpcogYw9/downloaded_packages


Old packages: 'devtools', 'pbdZMQ', 'R6', 'tidyr'
</code></pre>
</div>

<h1 id="exploratory-data-analysis">1. Exploratory data analysis</h1>

<p>We will be usig the Gene Expression dataset from <strong>Golub et al (1999)</strong>. The
gene expression data collected by Golub et al. (1999) are among the most
classical in bioinformatics. A selection of the set is called <code class="highlighter-rouge">golub</code> which is
contained in the <code class="highlighter-rouge">multtest</code> package loaded before.</p>

<p>The data consist of gene expression values of 3051 genes (rows) from 38 leukemia
patients Pre-processing was done as described in Dudoit et al. (2002). The R
code for pre-processing is available in the file ../doc/golub.R.</p>

<p><strong>Source</strong>:
Golub et al. (1999). Molecular classification of cancer: class discovery and
class prediction by gene expression monitoring, Science, Vol. 286:531-537.
http://www-genome.wi.mit.edu/MPR/ .</p>

<p><br /></p>
<font color="#00bcd4"> In [2]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">require</span><span class="p">(</span><span class="n">multtest</span><span class="p">);</span><span class="w">

</span><span class="c1"># Usage
</span><span class="n">data</span><span class="p">(</span><span class="n">golub</span><span class="p">);</span><span class="w">

</span><span class="c1"># If you need more information on the data set just
# uncomment the line below
</span><span class="err">#</span><span class="w"> </span><span class="o">?</span><span class="n">golub</span></code></pre></figure>

<div class="highlighter-rouge"><pre class="highlight"><code>Loading required package: multtest
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: ‘BiocGenerics’

The following objects are masked from ‘package:parallel’:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from ‘package:stats’:

    IQR, mad, xtabs

The following objects are masked from ‘package:base’:

    anyDuplicated, append, as.data.frame, cbind, colnames, do.call,
    duplicated, eval, evalq, Filter, Find, get, grep, grepl, intersect,
    is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
    paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
    Reduce, rownames, sapply, setdiff, sort, table, tapply, union,
    unique, unsplit, which, which.max, which.min

Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.
</code></pre>
</div>

<p>Data set values:
- <code class="highlighter-rouge">golub</code>: matrix of gene expression levels for the 38 tumor mRNA samples, rows
correspond to genes (3051 genes) and columns to mRNA samples.
- <code class="highlighter-rouge">golub.cl</code>: numeric vector indicating the tumor class, 27 acute lymphoblastic
leukemia (ALL) cases (code 0) and 11 acute myeloid leukemia (AML) cases (code
1).
- <code class="highlighter-rouge">golub.names</code>: a matrix containing the names of the 3051 genes for the
expression matrix golub. The three columns correspond to the gene index, ID, and
Name, respectively.</p>

<p><br /></p>
<font color="#00bcd4"> In [3]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Checking the dimension of the data
</span><span class="nf">dim</span><span class="p">(</span><span class="n">golub</span><span class="p">)</span></code></pre></figure>

<ol class="list-inline">
	<li>3051</li>
	<li>38</li>
</ol>

<p><br /></p>
<font color="#00bcd4"> In [4]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># we will have a look at the first rows contained in the data set
</span><span class="n">head</span><span class="p">(</span><span class="n">golub</span><span class="p">)</span></code></pre></figure>

<table>
<tbody>
	<tr><td>-1.45769</td><td>-1.39420</td><td>-1.42779</td><td>-1.40715</td><td>-1.42668</td><td>-1.21719</td><td>-1.37386</td><td>-1.36832</td><td>-1.47649</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td>-1.26183</td><td>-1.44434</td><td> 1.10147</td><td>-1.34158</td><td>-1.22961</td><td>-0.75919</td><td>0.84905 </td><td>-0.66465</td></tr>
	<tr><td>-0.75161</td><td>-1.26278</td><td>-0.09052</td><td>-0.99596</td><td>-1.24245</td><td>-0.69242</td><td>-1.37386</td><td>-0.50803</td><td>-1.04533</td><td>-0.81257</td><td>⋯       </td><td>-1.08902</td><td>-1.05094</td><td>-1.26183</td><td>-1.25918</td><td> 0.97813</td><td>-0.79357</td><td>-1.22961</td><td>-0.71792</td><td>0.45127 </td><td>-0.45804</td></tr>
	<tr><td> 0.45695</td><td>-0.09654</td><td> 0.90325</td><td>-0.07194</td><td> 0.03232</td><td> 0.09713</td><td>-0.11978</td><td> 0.23381</td><td> 0.23987</td><td> 0.44201</td><td>⋯       </td><td>-0.43377</td><td>-0.10823</td><td>-0.29385</td><td> 0.05067</td><td> 1.69430</td><td>-0.12472</td><td> 0.04609</td><td> 0.24347</td><td>0.90774 </td><td> 0.46509</td></tr>
	<tr><td> 3.13533</td><td> 0.21415</td><td> 2.08754</td><td> 2.23467</td><td> 0.93811</td><td> 2.24089</td><td> 3.36576</td><td> 1.97859</td><td> 2.66468</td><td>-1.21583</td><td>⋯       </td><td> 0.29598</td><td>-1.29865</td><td> 2.76869</td><td> 2.08960</td><td> 0.70003</td><td> 0.13854</td><td> 1.75908</td><td> 0.06151</td><td>1.30297 </td><td> 0.58186</td></tr>
	<tr><td> 2.76569</td><td>-1.27045</td><td> 1.60433</td><td> 1.53182</td><td> 1.63728</td><td> 1.85697</td><td> 3.01847</td><td> 1.12853</td><td> 2.17016</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td> 2.00518</td><td> 1.17454</td><td>-1.47218</td><td>-1.34158</td><td> 1.55086</td><td>-1.18107</td><td>1.01596 </td><td> 0.15788</td></tr>
	<tr><td> 2.64342</td><td> 1.01416</td><td> 1.70477</td><td> 1.63845</td><td>-0.36075</td><td> 1.73451</td><td> 3.36576</td><td> 0.96870</td><td> 2.72368</td><td>-1.21583</td><td>⋯       </td><td>-1.08902</td><td>-1.29865</td><td> 1.73780</td><td> 0.89347</td><td>-0.52883</td><td>-1.22168</td><td> 0.90832</td><td>-1.39906</td><td>0.51266 </td><td> 1.36249</td></tr>
</tbody>
</table>

<p>The gene names are collected in the matrix <code class="highlighter-rouge">golub.gnames</code> of which the columns
correspond to the gene index, ID, and Name, respectively.</p>

<p><br /></p>
<font color="#00bcd4"> In [5]: </font>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Adding 3051 gene names
</span><span class="n">row.names</span><span class="p">(</span><span class="n">golub</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">golub.gnames</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span><span class="w">

</span><span class="n">head</span><span class="p">(</span><span class="n">golub</span><span class="p">)</span></code></pre></figure>

<table>
<tbody>
	&lt;tr"&gt;<th scope="row">AFFX-HUMISGF3A/M97935_MA_at&lt;/th"&gt;&lt;td"&gt;-1.45769&lt;/td"&gt;&lt;td"&gt;-1.39420&lt;/td"&gt;&lt;td"&gt;-1.42779&lt;/td"&gt;&lt;td"&gt;-1.40715&lt;/td"&gt;&lt;td"&gt;-1.42668&lt;/td"&gt;&lt;td"&gt;-1.21719&lt;/td"&gt;&lt;td"&gt;-1.37386&lt;/td"&gt;&lt;td"&gt;-1.36832&lt;/td"&gt;&lt;td"&gt;-1.47649&lt;/td"&gt;&lt;td"&gt;-1.21583&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-1.08902&lt;/td"&gt;&lt;td"&gt;-1.29865&lt;/td"&gt;&lt;td"&gt;-1.26183&lt;/td"&gt;&lt;td"&gt;-1.44434&lt;/td"&gt;&lt;td"&gt; 1.10147&lt;/td"&gt;&lt;td"&gt;-1.34158&lt;/td"&gt;&lt;td"&gt;-1.22961&lt;/td"&gt;&lt;td"&gt;-0.75919&lt;/td"&gt;&lt;td"&gt;0.84905 &lt;/td"&gt;&lt;td"&gt;-0.66465&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMISGF3A/M97935_MB_at&lt;/th"&gt;&lt;td"&gt;-0.75161&lt;/td"&gt;&lt;td"&gt;-1.26278&lt;/td"&gt;&lt;td"&gt;-0.09052&lt;/td"&gt;&lt;td"&gt;-0.99596&lt;/td"&gt;&lt;td"&gt;-1.24245&lt;/td"&gt;&lt;td"&gt;-0.69242&lt;/td"&gt;&lt;td"&gt;-1.37386&lt;/td"&gt;&lt;td"&gt;-0.50803&lt;/td"&gt;&lt;td"&gt;-1.04533&lt;/td"&gt;&lt;td"&gt;-0.81257&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-1.08902&lt;/td"&gt;&lt;td"&gt;-1.05094&lt;/td"&gt;&lt;td"&gt;-1.26183&lt;/td"&gt;&lt;td"&gt;-1.25918&lt;/td"&gt;&lt;td"&gt; 0.97813&lt;/td"&gt;&lt;td"&gt;-0.79357&lt;/td"&gt;&lt;td"&gt;-1.22961&lt;/td"&gt;&lt;td"&gt;-0.71792&lt;/td"&gt;&lt;td"&gt;0.45127 &lt;/td"&gt;&lt;td"&gt;-0.45804&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMISGF3A/M97935_3_at&lt;/th"&gt;&lt;td"&gt; 0.45695&lt;/td"&gt;&lt;td"&gt;-0.09654&lt;/td"&gt;&lt;td"&gt; 0.90325&lt;/td"&gt;&lt;td"&gt;-0.07194&lt;/td"&gt;&lt;td"&gt; 0.03232&lt;/td"&gt;&lt;td"&gt; 0.09713&lt;/td"&gt;&lt;td"&gt;-0.11978&lt;/td"&gt;&lt;td"&gt; 0.23381&lt;/td"&gt;&lt;td"&gt; 0.23987&lt;/td"&gt;&lt;td"&gt; 0.44201&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-0.43377&lt;/td"&gt;&lt;td"&gt;-0.10823&lt;/td"&gt;&lt;td"&gt;-0.29385&lt;/td"&gt;&lt;td"&gt; 0.05067&lt;/td"&gt;&lt;td"&gt; 1.69430&lt;/td"&gt;&lt;td"&gt;-0.12472&lt;/td"&gt;&lt;td"&gt; 0.04609&lt;/td"&gt;&lt;td"&gt; 0.24347&lt;/td"&gt;&lt;td"&gt;0.90774 &lt;/td"&gt;&lt;td"&gt; 0.46509&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMRGE/M10098_5_at&lt;/th"&gt;&lt;td"&gt; 3.13533&lt;/td"&gt;&lt;td"&gt; 0.21415&lt;/td"&gt;&lt;td"&gt; 2.08754&lt;/td"&gt;&lt;td"&gt; 2.23467&lt;/td"&gt;&lt;td"&gt; 0.93811&lt;/td"&gt;&lt;td"&gt; 2.24089&lt;/td"&gt;&lt;td"&gt; 3.36576&lt;/td"&gt;&lt;td"&gt; 1.97859&lt;/td"&gt;&lt;td"&gt; 2.66468&lt;/td"&gt;&lt;td"&gt;-1.21583&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt; 0.29598&lt;/td"&gt;&lt;td"&gt;-1.29865&lt;/td"&gt;&lt;td"&gt; 2.76869&lt;/td"&gt;&lt;td"&gt; 2.08960&lt;/td"&gt;&lt;td"&gt; 0.70003&lt;/td"&gt;&lt;td"&gt; 0.13854&lt;/td"&gt;&lt;td"&gt; 1.75908&lt;/td"&gt;&lt;td"&gt; 0.06151&lt;/td"&gt;&lt;td"&gt;1.30297 &lt;/td"&gt;&lt;td"&gt; 0.58186&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMRGE/M10098_M_at&lt;/th"&gt;&lt;td"&gt; 2.76569&lt;/td"&gt;&lt;td"&gt;-1.27045&lt;/td"&gt;&lt;td"&gt; 1.60433&lt;/td"&gt;&lt;td"&gt; 1.53182&lt;/td"&gt;&lt;td"&gt; 1.63728&lt;/td"&gt;&lt;td"&gt; 1.85697&lt;/td"&gt;&lt;td"&gt; 3.01847&lt;/td"&gt;&lt;td"&gt; 1.12853&lt;/td"&gt;&lt;td"&gt; 2.17016&lt;/td"&gt;&lt;td"&gt;-1.21583&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-1.08902&lt;/td"&gt;&lt;td"&gt;-1.29865&lt;/td"&gt;&lt;td"&gt; 2.00518&lt;/td"&gt;&lt;td"&gt; 1.17454&lt;/td"&gt;&lt;td"&gt;-1.47218&lt;/td"&gt;&lt;td"&gt;-1.34158&lt;/td"&gt;&lt;td"&gt; 1.55086&lt;/td"&gt;&lt;td"&gt;-1.18107&lt;/td"&gt;&lt;td"&gt;1.01596 &lt;/td"&gt;&lt;td"&gt; 0.15788&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMRGE/M10098_3_at&lt;/th"&gt;&lt;td"&gt; 2.64342&lt;/td"&gt;&lt;td"&gt; 1.01416&lt;/td"&gt;&lt;td"&gt; 1.70477&lt;/td"&gt;&lt;td"&gt; 1.63845&lt;/td"&gt;&lt;td"&gt;-0.36075&lt;/td"&gt;&lt;td"&gt; 1.73451&lt;/td"&gt;&lt;td"&gt; 3.36576&lt;/td"&gt;&lt;td"&gt; 0.96870&lt;/td"&gt;&lt;td"&gt; 2.72368&lt;/td"&gt;&lt;td"&gt;-1.21583&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-1.08902&lt;/td"&gt;&lt;td"&gt;-1.29865&lt;/td"&gt;&lt;td"&gt; 1.73780&lt;/td"&gt;&lt;td"&gt; 0.89347&lt;/td"&gt;&lt;td"&gt;-0.52883&lt;/td"&gt;&lt;td"&gt;-1.22168&lt;/td"&gt;&lt;td"&gt; 0.90832&lt;/td"&gt;&lt;td"&gt;-1.39906&lt;/td"&gt;&lt;td"&gt;0.51266 &lt;/td"&gt;&lt;td"&gt; 1.36249&lt;/td"&gt;&lt;/tr"&gt;
&lt;/tbody&gt;
&lt;/table&gt;



<br />
<font color="#00bcd4"> In [7]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Let's just have a look at the top 20 genes ID's contained in golub.gnames
</span><span class="n">head</span><span class="p">(</span><span class="n">golub.gnames</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">)</span></code></pre></figure>



<ol class="list-inline">
	<li>'AFFX-HUMISGF3A/M97935_MA_at (endogenous control)'</li>
	<li>'AFFX-HUMISGF3A/M97935_MB_at (endogenous control)'</li>
	<li>'AFFX-HUMISGF3A/M97935_3_at (endogenous control)'</li>
	<li>'AFFX-HUMRGE/M10098_5_at (endogenous control)'</li>
	<li>'AFFX-HUMRGE/M10098_M_at (endogenous control)'</li>
	<li>'AFFX-HUMRGE/M10098_3_at (endogenous control)'</li>
	<li>'AFFX-HUMGAPDH/M33197_5_at (endogenous control)'</li>
	<li>'AFFX-HUMGAPDH/M33197_M_at (endogenous control)'</li>
	<li>'AFFX-HSAC07/X00351_5_at (endogenous control)'</li>
	<li>'AFFX-HSAC07/X00351_M_at (endogenous control)'</li>
	<li>'AFFX-HUMTFRR/M11507_5_at (endogenous control)'</li>
	<li>'AFFX-HUMTFRR/M11507_M_at (endogenous control)'</li>
	<li>'AFFX-HUMTFRR/M11507_3_at (endogenous control)'</li>
	<li>'AFFX-M27830_5_at (endogenous control)'</li>
	<li>'AFFX-M27830_M_at (endogenous control)'</li>
	<li>'AFFX-M27830_3_at (endogenous control)'</li>
	<li>'AFFX-HSAC07/X00351_3_st (endogenous control)'</li>
	<li>'AFFX-HUMGAPDH/M33197_M_st (endogenous control)'</li>
	<li>'AFFX-HUMGAPDH/M33197_3_st (endogenous control)'</li>
	<li>'AFFX-HSAC07/X00351_M_st (endogenous control)'</li>
</ol>



Twenty seven patients are diagnosed as acute lymphoblastic leukemia (ALL) and
eleven as acute myeloid leukemia (AML). The tumor class is given by the numeric
vector golub.cl, where ALL is indicated by 0 and AML by 1.

<br />
<font color="#00bcd4"> In [8]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">colnames</span><span class="p">(</span><span class="n">golub</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">golub.cl</span><span class="w">

</span><span class="n">head</span><span class="p">(</span><span class="n">golub</span><span class="p">)</span></code></pre></figure>



<table>
&lt;thead"&gt;&lt;tr"&gt;&lt;th"&gt;&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">⋯&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;&lt;/tr"&gt;&lt;/thead"&gt;
<tbody>
	&lt;tr"&gt;<th scope="row">AFFX-HUMISGF3A/M97935_MA_at&lt;/th"&gt;&lt;td"&gt;-1.45769&lt;/td"&gt;&lt;td"&gt;-1.39420&lt;/td"&gt;&lt;td"&gt;-1.42779&lt;/td"&gt;&lt;td"&gt;-1.40715&lt;/td"&gt;&lt;td"&gt;-1.42668&lt;/td"&gt;&lt;td"&gt;-1.21719&lt;/td"&gt;&lt;td"&gt;-1.37386&lt;/td"&gt;&lt;td"&gt;-1.36832&lt;/td"&gt;&lt;td"&gt;-1.47649&lt;/td"&gt;&lt;td"&gt;-1.21583&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-1.08902&lt;/td"&gt;&lt;td"&gt;-1.29865&lt;/td"&gt;&lt;td"&gt;-1.26183&lt;/td"&gt;&lt;td"&gt;-1.44434&lt;/td"&gt;&lt;td"&gt; 1.10147&lt;/td"&gt;&lt;td"&gt;-1.34158&lt;/td"&gt;&lt;td"&gt;-1.22961&lt;/td"&gt;&lt;td"&gt;-0.75919&lt;/td"&gt;&lt;td"&gt;0.84905 &lt;/td"&gt;&lt;td"&gt;-0.66465&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMISGF3A/M97935_MB_at&lt;/th"&gt;&lt;td"&gt;-0.75161&lt;/td"&gt;&lt;td"&gt;-1.26278&lt;/td"&gt;&lt;td"&gt;-0.09052&lt;/td"&gt;&lt;td"&gt;-0.99596&lt;/td"&gt;&lt;td"&gt;-1.24245&lt;/td"&gt;&lt;td"&gt;-0.69242&lt;/td"&gt;&lt;td"&gt;-1.37386&lt;/td"&gt;&lt;td"&gt;-0.50803&lt;/td"&gt;&lt;td"&gt;-1.04533&lt;/td"&gt;&lt;td"&gt;-0.81257&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-1.08902&lt;/td"&gt;&lt;td"&gt;-1.05094&lt;/td"&gt;&lt;td"&gt;-1.26183&lt;/td"&gt;&lt;td"&gt;-1.25918&lt;/td"&gt;&lt;td"&gt; 0.97813&lt;/td"&gt;&lt;td"&gt;-0.79357&lt;/td"&gt;&lt;td"&gt;-1.22961&lt;/td"&gt;&lt;td"&gt;-0.71792&lt;/td"&gt;&lt;td"&gt;0.45127 &lt;/td"&gt;&lt;td"&gt;-0.45804&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMISGF3A/M97935_3_at&lt;/th"&gt;&lt;td"&gt; 0.45695&lt;/td"&gt;&lt;td"&gt;-0.09654&lt;/td"&gt;&lt;td"&gt; 0.90325&lt;/td"&gt;&lt;td"&gt;-0.07194&lt;/td"&gt;&lt;td"&gt; 0.03232&lt;/td"&gt;&lt;td"&gt; 0.09713&lt;/td"&gt;&lt;td"&gt;-0.11978&lt;/td"&gt;&lt;td"&gt; 0.23381&lt;/td"&gt;&lt;td"&gt; 0.23987&lt;/td"&gt;&lt;td"&gt; 0.44201&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-0.43377&lt;/td"&gt;&lt;td"&gt;-0.10823&lt;/td"&gt;&lt;td"&gt;-0.29385&lt;/td"&gt;&lt;td"&gt; 0.05067&lt;/td"&gt;&lt;td"&gt; 1.69430&lt;/td"&gt;&lt;td"&gt;-0.12472&lt;/td"&gt;&lt;td"&gt; 0.04609&lt;/td"&gt;&lt;td"&gt; 0.24347&lt;/td"&gt;&lt;td"&gt;0.90774 &lt;/td"&gt;&lt;td"&gt; 0.46509&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMRGE/M10098_5_at&lt;/th"&gt;&lt;td"&gt; 3.13533&lt;/td"&gt;&lt;td"&gt; 0.21415&lt;/td"&gt;&lt;td"&gt; 2.08754&lt;/td"&gt;&lt;td"&gt; 2.23467&lt;/td"&gt;&lt;td"&gt; 0.93811&lt;/td"&gt;&lt;td"&gt; 2.24089&lt;/td"&gt;&lt;td"&gt; 3.36576&lt;/td"&gt;&lt;td"&gt; 1.97859&lt;/td"&gt;&lt;td"&gt; 2.66468&lt;/td"&gt;&lt;td"&gt;-1.21583&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt; 0.29598&lt;/td"&gt;&lt;td"&gt;-1.29865&lt;/td"&gt;&lt;td"&gt; 2.76869&lt;/td"&gt;&lt;td"&gt; 2.08960&lt;/td"&gt;&lt;td"&gt; 0.70003&lt;/td"&gt;&lt;td"&gt; 0.13854&lt;/td"&gt;&lt;td"&gt; 1.75908&lt;/td"&gt;&lt;td"&gt; 0.06151&lt;/td"&gt;&lt;td"&gt;1.30297 &lt;/td"&gt;&lt;td"&gt; 0.58186&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMRGE/M10098_M_at&lt;/th"&gt;&lt;td"&gt; 2.76569&lt;/td"&gt;&lt;td"&gt;-1.27045&lt;/td"&gt;&lt;td"&gt; 1.60433&lt;/td"&gt;&lt;td"&gt; 1.53182&lt;/td"&gt;&lt;td"&gt; 1.63728&lt;/td"&gt;&lt;td"&gt; 1.85697&lt;/td"&gt;&lt;td"&gt; 3.01847&lt;/td"&gt;&lt;td"&gt; 1.12853&lt;/td"&gt;&lt;td"&gt; 2.17016&lt;/td"&gt;&lt;td"&gt;-1.21583&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-1.08902&lt;/td"&gt;&lt;td"&gt;-1.29865&lt;/td"&gt;&lt;td"&gt; 2.00518&lt;/td"&gt;&lt;td"&gt; 1.17454&lt;/td"&gt;&lt;td"&gt;-1.47218&lt;/td"&gt;&lt;td"&gt;-1.34158&lt;/td"&gt;&lt;td"&gt; 1.55086&lt;/td"&gt;&lt;td"&gt;-1.18107&lt;/td"&gt;&lt;td"&gt;1.01596 &lt;/td"&gt;&lt;td"&gt; 0.15788&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">AFFX-HUMRGE/M10098_3_at&lt;/th"&gt;&lt;td"&gt; 2.64342&lt;/td"&gt;&lt;td"&gt; 1.01416&lt;/td"&gt;&lt;td"&gt; 1.70477&lt;/td"&gt;&lt;td"&gt; 1.63845&lt;/td"&gt;&lt;td"&gt;-0.36075&lt;/td"&gt;&lt;td"&gt; 1.73451&lt;/td"&gt;&lt;td"&gt; 3.36576&lt;/td"&gt;&lt;td"&gt; 0.96870&lt;/td"&gt;&lt;td"&gt; 2.72368&lt;/td"&gt;&lt;td"&gt;-1.21583&lt;/td"&gt;&lt;td"&gt;⋯       &lt;/td"&gt;&lt;td"&gt;-1.08902&lt;/td"&gt;&lt;td"&gt;-1.29865&lt;/td"&gt;&lt;td"&gt; 1.73780&lt;/td"&gt;&lt;td"&gt; 0.89347&lt;/td"&gt;&lt;td"&gt;-0.52883&lt;/td"&gt;&lt;td"&gt;-1.22168&lt;/td"&gt;&lt;td"&gt; 0.90832&lt;/td"&gt;&lt;td"&gt;-1.39906&lt;/td"&gt;&lt;td"&gt;0.51266 &lt;/td"&gt;&lt;td"&gt; 1.36249&lt;/td"&gt;&lt;/tr"&gt;
&lt;/tbody&gt;
&lt;/table&gt;



Note that sometimes it is better to construct a factor which indicates the tumor
class of the patients. Such a factor could be used for instance to separate the
tumor groups for plotting purposes.  The factor (`gol.fac`) can be contructed as
follows.

<br />
<font color="#00bcd4"> In [50]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">gol.fac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">  </span><span class="n">factor</span><span class="p">(</span><span class="n">golub.cl</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="o">:</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"AML"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ALL"</span><span class="p">))</span></code></pre></figure>


Obviously, the labels corre-
spond to the two tumor classes. The evaluation of gol.fac=="ALL" returns
TRUE for the ¯rst twenty seven values and FALSE for the remaining eleven,
which is useful as a column index for selecting the expression values of the
ALL patients. The expression values of gene CCND3 Cyclin D3 from the
ALL patients can now be printed to the screen, as follows.

<br />
<font color="#00bcd4"> In [51]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="n">gol.fac</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"ALL"</span><span class="p">]</span></code></pre></figure>



<dl class="dl-horizontal">
	<dt>1</dt>
		<dd>0.88941</dd>
	<dt>1</dt>
		<dd>1.45014</dd>
	<dt>1</dt>
		<dd>0.42904</dd>
	<dt>1</dt>
		<dd>0.82667</dd>
	<dt>1</dt>
		<dd>0.63637</dd>
	<dt>1</dt>
		<dd>1.0225</dd>
	<dt>1</dt>
		<dd>0.12758</dd>
	<dt>1</dt>
		<dd>-0.74333</dd>
	<dt>1</dt>
		<dd>0.73784</dd>
	<dt>1</dt>
		<dd>0.4947</dd>
	<dt>1</dt>
		<dd>1.12058</dd>
</dl>



## Creating the exploratory plots

### 1.1\. Plotting the value of gene (CCND3) in all nRNA samples (M92287_at)

We shall first have a look at the expression values of a gener with manufacurer
name `M92278_at`, which is known in biology as "CCND3 Cyclin D3".

The expression values of this gene are collected in row 1042 of golub. To load
the data and to obtain the relevant information from row 1042 of golub.gnames,
use the following:

<br />
<font color="#00bcd4"> In [8]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">mygene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="p">]</span></code></pre></figure>


The data has now been stored in the `golub` matrix. We will now plot the
expression values od the gene CCND3 Cyclin D3.

<br />
<font color="#00bcd4"> In [9]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">plot</span><span class="p">(</span><span class="n">mygene</span><span class="p">)</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_21_0.png)


In the previous plot we just used the default plotting preferences within R base
plotting.We can do some improvements so that the plot is easily understood.

<br />
<font color="#00bcd4"> In [37]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">plot</span><span class="p">(</span><span class="n">mygene</span><span class="p">,</span><span class="w"> </span><span class="n">pch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'slateblue'</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Expression value of gene: CCND3'</span><span class="p">,</span><span class="w"> 
    </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">' Gene expression values of CCND3 Cyclin D3'</span><span class="p">)</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_23_0.png)


In this plot the vertical axis corresponds to the size of the expression values
and the horizontal axis the index of the patients.

### 1.2\. Gene expression between patient 1 (ALL) and patient 38 (AML)

<br />
<font color="#00bcd4"> In [11]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">plot</span><span class="p">(</span><span class="n">golub</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">golub</span><span class="p">[,</span><span class="m">38</span><span class="p">])</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_26_0.png)


Adding diagonal lines to the plot and changing axes labels


<br />
<font color="#00bcd4"> In [12]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">plot</span><span class="p">(</span><span class="n">golub</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">golub</span><span class="p">[,</span><span class="m">38</span><span class="p">],</span><span class="w"> </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Patient 1 (ALL)'</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Patient 38 (AML)'</span><span class="p">)</span><span class="w"> 
</span><span class="n">abline</span><span class="p">(</span><span class="n">a</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'mediumpurple'</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="w"> </span><span class="o">=</span><span class="m">3</span><span class="p">)</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_28_0.png)


### 1.3\. Scatter plots to detect independence


<br />
<font color="#00bcd4"> In [13]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">mysamplist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">golub</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">15</span><span class="p">)]</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">mysamplist</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">15</span><span class="p">)</span></code></pre></figure>


<br />
<font color="#00bcd4"> In [14]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">plot</span><span class="p">(</span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">mysamplist</span><span class="p">),</span><span class="w"> </span><span class="n">pch</span><span class="o">=</span><span class="s1">'.'</span><span class="p">)</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_31_0.png)


### 1.4\. Bar plot of 4 cyclin genes expression values in 3 ALL and AML patients

We will analyse the expression values of the `D13639_at, M92287_at, U11791_at,
Z36714_AT` genes in three chosen AML and ALL patients

<br />
<font color="#00bcd4"> In [15]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">mygenelist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">golub</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">85</span><span class="p">,</span><span class="w"> </span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="m">1212</span><span class="p">,</span><span class="w"> </span><span class="m">2240</span><span class="p">),</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="m">36</span><span class="o">:</span><span class="m">38</span><span class="p">)]</span><span class="w">

</span><span class="c1"># having a look at the data set chosen
</span><span class="n">mygenelist</span></code></pre></figure>



<table>
&lt;thead"&gt;&lt;tr"&gt;&lt;th"&gt;&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">0&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;<th scope="col">1&lt;/th"&gt;&lt;/tr"&gt;&lt;/thead"&gt;
<tbody>
	&lt;tr"&gt;<th scope="row">D13639_at&lt;/th"&gt;&lt;td"&gt; 2.09511&lt;/td"&gt;&lt;td"&gt; 1.71953&lt;/td"&gt;&lt;td"&gt;-1.46227&lt;/td"&gt;&lt;td"&gt;-0.92935&lt;/td"&gt;&lt;td"&gt;-0.11091&lt;/td"&gt;&lt;td"&gt; 1.15591&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">M92287_at&lt;/th"&gt;&lt;td"&gt; 2.10892&lt;/td"&gt;&lt;td"&gt; 1.52405&lt;/td"&gt;&lt;td"&gt; 1.96403&lt;/td"&gt;&lt;td"&gt; 0.73784&lt;/td"&gt;&lt;td"&gt; 0.49470&lt;/td"&gt;&lt;td"&gt; 1.12058&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">U11791_at&lt;/th"&gt;&lt;td"&gt;-0.11439&lt;/td"&gt;&lt;td"&gt;-0.72887&lt;/td"&gt;&lt;td"&gt;-0.39674&lt;/td"&gt;&lt;td"&gt;-0.94364&lt;/td"&gt;&lt;td"&gt; 0.05047&lt;/td"&gt;&lt;td"&gt; 0.05905&lt;/td"&gt;&lt;/tr"&gt;
	&lt;tr"&gt;<th scope="row">Z36714_at&lt;/th"&gt;&lt;td"&gt;-1.45769&lt;/td"&gt;&lt;td"&gt;-1.39420&lt;/td"&gt;&lt;td"&gt;-1.46227&lt;/td"&gt;&lt;td"&gt;-1.39906&lt;/td"&gt;&lt;td"&gt;-1.34579&lt;/td"&gt;&lt;td"&gt;-1.32403&lt;/td"&gt;&lt;/tr"&gt;
&lt;/tbody&gt;
&lt;/table&gt;



<br />
<font color="#00bcd4"> In [26]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">barplot</span><span class="p">(</span><span class="n">mygenelist</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_34_0.png)


The plot is not very easy to read, so we will add some colurs and a legend so we
know which gene each bar segment corresponds to.


<br />
<font color="#00bcd4"> In [28]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># custom colours 
</span><span class="n">colours</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'lightblue'</span><span class="p">,</span><span class="w">  </span><span class="s1">'mediumpurple'</span><span class="p">,</span><span class="w"> </span><span class="s1">'lavender'</span><span class="p">,</span><span class="w"> </span><span class="s1">'midnightblue'</span><span class="p">)</span><span class="w">

</span><span class="n">barplot</span><span class="p">(</span><span class="n">mygenelist</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colours</span><span class="p">,</span><span class="w"> </span><span class="n">legend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_36_0.png)


In this case the patients are indicated on the `X` axis (0 and 1 respectively)
while the gene expression level is indicate on the `Y` axis.

We can make some improvements to the plots.
Let's have a look at the `barplot` arguments:

<br />
<font color="#00bcd4"> In [17]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="o">?</span><span class="n">barplot</span></code></pre></figure>


We are going to focus on only a few of the histgram arguments:
- `beside`: `TRUE` for the bars to be displayed as justapoxed bars, `FALSE` for
stacked bars
- `horiz` : `FALSE` bars displayed vertically with the first bar to the left,
`TRUE` bars are displayed horizontally with the first at the bottom.
- `ylim`, `xlim` :  limits for the y and x axes
- `col`: colour choices

<br />
<font color="#00bcd4"> In [35]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">barplot</span><span class="p">(</span><span class="n">mygenelist</span><span class="p">,</span><span class="w"> </span><span class="n">horiz</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colours</span><span class="p">,</span><span class="w"> </span><span class="n">legend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w">
       </span><span class="n">ylab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Patient'</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Gene expression level'</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="s1">'Cycline genes expression'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_40_0.png)


In the plot above we presented the barplots horizontally and added some colours,
which makes it easier to understand the data presented.
You can also use the barplots to represent the mean and standard error which we
will be doing in the following sections.

### 1.5\. Plotting the mean

In the following we will compute the mean for the expression values of both the
ALL and AML patients. We will be using the same 4 cycline genes used in the
example above.

First we will compute the ALL and AML for all the patients. Once the means are
computed they are combined into asingle data frame.

Finally, the means are plotted using the `barplot` function.

<br />
<font color="#00bcd4"> In [36]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Calculating the mean of the chosen genes from patient 1 to 27 and 28 to 38
</span><span class="n">ALLmean</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowMeans</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">85</span><span class="p">,</span><span class="m">1042</span><span class="p">,</span><span class="m">1212</span><span class="p">,</span><span class="m">2240</span><span class="p">),</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">27</span><span class="p">)])</span><span class="w">
</span><span class="n">AMLmean</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowMeans</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">85</span><span class="p">,</span><span class="m">1042</span><span class="p">,</span><span class="m">1212</span><span class="p">,</span><span class="m">2240</span><span class="p">),</span><span class="nf">c</span><span class="p">(</span><span class="m">28</span><span class="o">:</span><span class="m">38</span><span class="p">)])</span><span class="w">

</span><span class="c1"># Combining the mean matrices previously calculated
</span><span class="n">dataheight</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="n">ALLmean</span><span class="p">,</span><span class="w"> </span><span class="n">AMLmean</span><span class="p">)</span><span class="w">

</span><span class="c1"># Plotting 
</span><span class="n">barx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">barplot</span><span class="p">(</span><span class="n">dataheight</span><span class="p">,</span><span class="w"> </span><span class="n">beside</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">horiz</span><span class="o">=</span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="w"> </span><span class="n">colours</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-2</span><span class="p">,</span><span class="m">2.5</span><span class="p">),</span><span class="w"> 
                </span><span class="n">legend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Gene expression level'</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Cycline genes mean expression
in AML and ALL patients'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_43_0.png)


### 1.6\. Adding error bars to the previous plot


In the previous section we computed the mean expression level for 4 cycline
genes between the AML and ALL patients. Sometimes it is useful to add error bars
to the plots (as the one above) to convey the uncertainty in the data presented.

For such a purpose we often use the **Standard Deviation**:


$$ \sigma = \sqrt{\frac{\sum_{i=1}^{n}\left(x_i -\mu \right)^2}{N}}$$


which in turn tells us how much the values in a certain group tend to deviate
from their mean value.

Let's start calculating the Standard Deviation of the data.



<br />
<font color="#00bcd4"> In [54]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Calculating the SD
</span><span class="n">ALLsd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">85</span><span class="p">,</span><span class="m">1042</span><span class="p">,</span><span class="m">1212</span><span class="p">,</span><span class="m">2240</span><span class="p">),</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">27</span><span class="p">)],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="p">)</span><span class="w">
</span><span class="n">nALL</span><span class="o">=</span><span class="nf">length</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">27</span><span class="p">))</span><span class="w">
</span><span class="n">AMLsd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">85</span><span class="p">,</span><span class="m">1042</span><span class="p">,</span><span class="m">1212</span><span class="p">,</span><span class="m">2240</span><span class="p">),</span><span class="nf">c</span><span class="p">(</span><span class="m">28</span><span class="o">:</span><span class="m">38</span><span class="p">)],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="p">)</span><span class="w">
</span><span class="n">nAML</span><span class="o">=</span><span class="nf">length</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">28</span><span class="o">:</span><span class="m">38</span><span class="p">))</span><span class="w">

</span><span class="c1"># Combining the data
</span><span class="n">datasd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="n">ALLsd</span><span class="p">,</span><span class="w"> </span><span class="n">AMLsd</span><span class="p">)</span></code></pre></figure>


Another measure used to quantify the deviation is the **standard error**, which
qutifies the variability in the **_means_** of our groups instead of reporting
the variability among the data points.

A relatively straigtforward way to compute this is by assuming if we were to
repeat a  given experiment many many times, then it would roughly follow a
normal distribution. **Note – this is a big assumption**.  hence, if we assuemt
hat the means follow a nosmal distribution, then the standard error (_a.k.a.
variability of group means_) can be defined as:

$$ SE  = \frac{SD}{\sqrt{n}} $$

which in layman terms can be read as  “take the general variability of the
points around their group means (the standard deviation), and scale this number
by the number of points that you’ve collected”.

Since we have already computed the SD we can now compute the standard error
(SE).

<br />
<font color="#00bcd4"> In [55]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">datase</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="n">ALLsd</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="n">nALL</span><span class="p">),</span><span class="w"> </span><span class="n">AMLsd</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="n">nAML</span><span class="p">))</span></code></pre></figure>


Now we can create a plot of the mean data as well as the SE and SD.

<br />
<font color="#00bcd4"> In [63]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># creating a panel of 2 plots displayed in 1 row
</span><span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">))</span><span class="w">

</span><span class="c1"># Plot with the SD
</span><span class="n">datasdend</span><span class="o">&lt;-</span><span class="nf">abs</span><span class="p">(</span><span class="n">dataheight</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">datasd</span><span class="p">)</span><span class="w">
</span><span class="n">datasdend</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">),]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">datasdend</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">),]</span><span class="w">
</span><span class="n">barx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">barplot</span><span class="p">(</span><span class="n">dataheight</span><span class="p">,</span><span class="w"> </span><span class="n">beside</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">horiz</span><span class="o">=</span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colours</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-2</span><span class="p">,</span><span class="m">2.5</span><span class="p">),</span><span class="w">
               </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Data +  SD'</span><span class="p">)</span><span class="w">

</span><span class="n">arrows</span><span class="p">(</span><span class="n">barx</span><span class="p">,</span><span class="w"> </span><span class="n">dataheight</span><span class="p">,</span><span class="w"> </span><span class="n">barx</span><span class="p">,</span><span class="w"> </span><span class="n">datasdend</span><span class="p">,</span><span class="w"> </span><span class="n">angle</span><span class="o">=</span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.15</span><span class="p">,</span><span class="w"> 
       </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'magenta'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span><span class="w">

</span><span class="c1"># Plot with the se: error associated to the mean!
</span><span class="n">datasdend</span><span class="o">&lt;-</span><span class="nf">abs</span><span class="p">(</span><span class="n">dataheight</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">datase</span><span class="p">)</span><span class="w">
</span><span class="n">datasdend</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">),]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">-</span><span class="n">datasdend</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">),]</span><span class="w">
</span><span class="n">barx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">barplot</span><span class="p">(</span><span class="n">dataheight</span><span class="p">,</span><span class="w"> </span><span class="n">beside</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">horiz</span><span class="o">=</span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colours</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-2</span><span class="p">,</span><span class="m">2.5</span><span class="p">),</span><span class="w">
               </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Data + SE'</span><span class="p">)</span><span class="w">

</span><span class="n">arrows</span><span class="p">(</span><span class="n">barx</span><span class="p">,</span><span class="w"> </span><span class="n">dataheight</span><span class="p">,</span><span class="w"> </span><span class="n">barx</span><span class="p">,</span><span class="w"> </span><span class="n">datasdend</span><span class="p">,</span><span class="w"> </span><span class="n">angle</span><span class="o">=</span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.15</span><span class="p">,</span><span class="w">
       </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'magenta'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_49_0.png)


Note that the error bars for the SE are smaller than those for the SD. This is
no coincidence! As we increase N (in the SE equation), we will decrease the
error. Hence the standard error will **always** be smaller than the SD.

## 2. Data representation
This section presents some essential manners to display and visualize  data.

### 2.1 Frequency table
Discrete data occur when the values naturally fall into categories. A frequency
table simply gives the number of occurrences within a category.

A gene consists of a sequence of nucleotides (A; C; G; T)

The number of each nucleotide can be displayed in a frequency table.

This will be illustrated by the Zyxin gene which plays an important role in cell
adhesion The accession number (X94991.1) of one of its variants can be found in
a data base like NCBI (UniGene). The code below illustrates how to read the
sequence ”X94991.1” of the species homo sapiens from GenBank, to construct a
pie from a frequency table of the four nucleotides .

<br />
<font color="#00bcd4"> In [22]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">library</span><span class="p">(</span><span class="s1">'ape'</span><span class="p">)</span></code></pre></figure>


<br />
<font color="#00bcd4"> In [65]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">v</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">read.GenBank</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"X94991.1"</span><span class="p">),</span><span class="n">as.character</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">pie</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">X</span><span class="m">94991.1</span><span class="p">),</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colours</span><span class="p">)</span><span class="w">

</span><span class="c1"># prints the data as a table 
</span><span class="n">table</span><span class="p">(</span><span class="n">read.GenBank</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"X94991.1"</span><span class="p">),</span><span class="n">as.character</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span></code></pre></figure>



    
      a   c   g   t 
    410 789 573 394 



![png](http://localhost:4000/BAD_days/notebooks/tutorial_files/tutorial_53_1.png)


### 2.2 Stripcharts

An elementary method to visualize data is by using a so-called stripchart,
by which the values of the data are represented as e.g. small boxes
it is useful in combination with a factor that distinguishes members from
different experimental conditions or patients groups.

Once again we use the CCND3 Cyclin D3 data to generate the plots.

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># data(golub, package = "multtest")
</span><span class="n">gol.fac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">golub.cl</span><span class="p">,</span><span class="n">levels</span><span class="o">=</span><span class="m">0</span><span class="o">:</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">labels</span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"ALL"</span><span class="p">,</span><span class="s2">"AML"</span><span class="p">))</span><span class="w">

</span><span class="n">stripchart</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,]</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">gol.fac</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"jitter"</span><span class="p">,</span><span class="w"> 
           </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'slateblue'</span><span class="p">,</span><span class="w"> </span><span class="s1">'darkgrey'</span><span class="p">),</span><span class="w"> </span><span class="n">pch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">16</span><span class="p">)</span></code></pre></figure>


To produce two adjacent stripcharts one for the ALL and one for the AML
patients, we use the factor called `gol.fac`. From the above figure, it can be
observed that the CCND3 Cyclin D3 expression values of the ALL patients tend to
have larger expression values than those of the AML patient.

### 2.3 Histograms

Another method to visualize data is by dividing the range of data values into
a number of intervals and to plot the frequency per interval as a bar. Such
a plot is called a histogram.

We will now generate a histogram of the expression values of gene CCND

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">2</span><span class="p">))</span><span class="w">

</span><span class="n">hist</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="n">gol.fac</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"ALL"</span><span class="p">],</span><span class="w"> 
     </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'lightblue'</span><span class="p">,</span><span class="w"> </span><span class="n">border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'white'</span><span class="p">,</span><span class="w">
    </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Golub[1042], ALL'</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'ALL'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span><span class="w">

</span><span class="n">hist</span><span class="p">(</span><span class="n">golub</span><span class="p">,</span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> 
    </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'slateblue'</span><span class="p">,</span><span class="w"> </span><span class="n">border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'white'</span><span class="p">,</span><span class="w">
    </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w">  </span><span class="s1">'Golub'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span><span class="w">

</span><span class="n">hist</span><span class="p">(</span><span class="n">golub</span><span class="p">[,</span><span class="w"> </span><span class="n">gol.fac</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"AML"</span><span class="p">],</span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> 
     </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'slateblue'</span><span class="p">,</span><span class="w"> </span><span class="n">border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'white'</span><span class="p">,</span><span class="w">
    </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Golub, AML'</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'AML'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span><span class="w">

</span><span class="n">hist</span><span class="p">(</span><span class="n">golub</span><span class="p">[,</span><span class="w"> </span><span class="n">gol.fac</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"ALL"</span><span class="p">],</span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w">
     </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'slateblue'</span><span class="p">,</span><span class="w"> </span><span class="n">border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'white'</span><span class="p">,</span><span class="w">
    </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'Golub, ALL'</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'ALL'</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>


### 2.3 Boxplots

A popular method to display data is by
drawing a box around the Ørst and the third quartile (a bold line segment
for the median), and the smaller line segments (whiskers) for the smallest and
the largest data values. Such a data display is known as a box-and-whisker
plot

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Sort the values of one gene
</span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sort</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="n">gol.fac</span><span class="o">==</span><span class="s2">"ALL"</span><span class="p">],</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">x</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span></code></pre></figure>


A view on the distribution of the gene expression values of the `ALL` and `AML`
patients on gene CCND3 Cyclin D3 can be obtained by construction two separate
boxplots adjacent to each other:

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Even though we are creating two boxplots we only need one major graph
</span><span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">))</span><span class="w">
</span><span class="n">boxplot</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,]</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">gol.fac</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'lightblue'</span><span class="p">,</span><span class="w"> </span><span class="s1">'mediumpurple'</span><span class="p">))</span></code></pre></figure>


It can be observed that the gene expression values for ALL are larger than those
for AML. Furthermore, since the two sub-boxes around the median are more or less
equally wide, the data are quite symmetrically distributed around the median.

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">hist</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,],</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="w"> </span><span class="s1">'lightblue'</span><span class="p">,</span><span class="w"> </span><span class="n">border</span><span class="o">=</span><span class="w"> </span><span class="s1">'black'</span><span class="p">,</span><span class="w"> </span><span class="n">breaks</span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="n">freq</span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w">
</span><span class="n">lines</span><span class="p">(</span><span class="n">density</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,]),</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="w"> </span><span class="s1">'slateblue'</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w">
</span><span class="n">box</span><span class="p">()</span></code></pre></figure>


Now we can observe the distribution of all gene expressions values in all 38
patients

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">boxplot</span><span class="p">(</span><span class="n">golub</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="w"> </span><span class="s1">'lightblue'</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">border</span><span class="o">=</span><span class="s2">"black"</span><span class="p">,</span><span class="w"> </span><span class="n">pch</span><span class="o">=</span><span class="m">18</span><span class="p">)</span></code></pre></figure>


To compute exact values for the quartiles we need a sequence running from 0 to 1
with increments in steps of 0.25

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">pvec</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">seq</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0.25</span><span class="p">)</span><span class="w">
</span><span class="n">quantile</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="n">gol.fac</span><span class="o">==</span><span class="s1">'ALL'</span><span class="p">],</span><span class="w"> </span><span class="n">pvec</span><span class="p">)</span></code></pre></figure>


Outliers are data points lying far apart from the pattern set by the majority of
the data values. The implementation in R of the boxplot draws such outliers as
smalle circles.

A data point `x` is defined (graphically, not statistically) as an outlier point
if $$x &lt; 0.25 x -1.5\left(0.75 x -0.25 x\right) [x&gt;0.25x &gt;1.5(0.75x-0.25x)]$$


### 2.4 Q-Q plots

A method to visualize the distribution of gene expression values is y the so-
called quantile-quantile (Q-Q) plots. In such a plot the quantiles of the gene
expression values are displayed against the corresponding quantiles of the
normal distribution (bell-shaped).


To produce a Q-Q plot of the ALL gene expression values of CCND3 Cyclin D3 one
may use the following.

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">qqnorm</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="n">gol.fac</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s1">'ALL'</span><span class="p">])</span><span class="w">
</span><span class="n">qqline</span><span class="p">(</span><span class="n">golub</span><span class="p">[</span><span class="m">1042</span><span class="p">,</span><span class="w"> </span><span class="n">gol.fac</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s1">'ALL'</span><span class="p">],</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'mediumpurple'</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span></code></pre></figure>


It can be seen that most of the data points are on or near the straight line,
while a few others are further away. The above example illustrates a case where
the degree of non-normality is moderate so that a clear conclusion cannot be
drawn.


## 3. Loading tab-delimited data

<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">mydata</span><span class="o">&lt;-</span><span class="n">read.delim</span><span class="p">(</span><span class="s2">"./NeuralStemCellData.tab.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="nb">T</span><span class="p">)</span></code></pre></figure>


<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="nf">class</span><span class="p">(</span><span class="n">mydata</span><span class="p">)</span></code></pre></figure>


### Now try and do some exploratory analysis of your own on this data!


GvHD flow cytometry data

Only exract the CD3 positive cells


<br />
<font color="#00bcd4"> In [None]: </font>


<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">cor</span><span class="p">(</span><span class="n">mydata</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="n">mydata</span><span class="p">[,</span><span class="m">2</span><span class="p">])</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">mydata</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="n">mydata</span><span class="p">[,</span><span class="m">3</span><span class="p">])</span></code></pre></figure>


<br />
<font color="#00bcd4"> In [None]: </font>


</th></th></th></th></tbody></th></th></th></th></th></th></table></th></th></th></th></th></th></tbody></th></th></th></th></th></th></th></th></th></th></th></th></th></th></th></th></th></th></th></th></th></table></th></th></th></th></th></th></tbody></table>

        </main>
      </div>

      <div class="col-sm-1">
      </div>
      <!-- TOC sidebar, which will move to the top on a small screen -->
      <div class="col-sm-3 container-toc">
        <nav id="toc" data-spy="affix" data-toggle="toc"></nav>
      </div>
    </div>
  </div>
</div>

<label for="sidebar-checkbox" class="sidebar-toggle"></label>

</body>


</html>
